DataDog/datadog-agent
# [system-probe] Handle self-connect connection expiry

### What does this PR do?

This PR addresses a connection "leak" in the network code where it was not removing expired connections that are essentially duplicates of one another, with the same 5-tuple but different pids. This can happen if a process forks after connecting to itself; see the fork.py script in this PR. Since conntrack only tracks the 5-tuple even if the forked child exits, its connection still matches the 5-tuple in conntrack and se we never expire the child's connection as long as the parent is still alive. We now also check if the pid in a connection is dead to mark it expired if it is.

### Motivation

See above.

https://datadoghq.atlassian.net/browse/NET-985

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

1. Run the system-probe
2. Run https://github.com/DataDog/datadog-agent/blob/b1e15651e95797d48ef0cc715f8df069a941030b/pkg/network/tracer/testdata/fork.py . This will fork a child while the parent will wait (i.e. the script will not return). Keep this running in the background or switch to another terminal for the rest of the test
3. Run `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq '.conns[] | select(.laddr.port==33333 and .raddr.port==33333) '`. You should see two connections
4. Wait a little over 2 minutes (timeout for tcp connections)
5. Run `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq '.conns[] | select(.laddr.port==33333 and .raddr.port==33333) '`. You should see _one_ connection
